### ISSUE TITLE
Build a Confidential Single-Price Auction for Tokens with Sealed Bids using Zama's fhEVM

### ISSUE THREAD

**Issue by zaccherinij** [zama-ai/bounty-program#136]
# Overview

Create a Single Price auction for a quantity of tokens with sealed bids using Solidity and the FHEVM. The single price should be the lowest price from the bid required to purchase the last token sold.

## Description

An auction owner can create/launch a **Single-Price Auction** for a quantity of assets (e.g. ERC20) to be sold using either ether or an ERC20 token. 

Each participant, who wants to buy assets, can place **an encrypted bid** (the number of tokens to purchase and the price that each participant is willing to pay). **The settlement price is the price at which the last token is bought**. This price is paid by everyone regardless on the initial bids.

If the auction does not sell fully (e.g. no participant, not enough participant), it is up to the developer to decide what should be the best resolution mechanism (e.g. refund mechanism, execution at the lowest price). 

Other corner/edge cases may include:
- whether an address can participate more than once
- whether a user can modify an existing bid before the auction end
- how long the auction should last
- how to determine the proper resolution if the two lowest bids are equal
- how to prevent participation in an auction without the funds (i.e. the locking mechanism)

> [!TIP]
> Design choices should be clearly documented. All corner/edge cases must be handled by the developer.

## Example

**Initialization:** Alice creates an auction to sell 1,000,000 tokens in ether. 

**Bidding:** Bob bids 0.000002 ether per token for 500,000 tokens, Carol bids 0.000008 ether per token for 600,000 tokens, David bids 0.00000000001 per token for 1,000,000 tokens.

**Resolution:** Carol gets 600,000 tokens, Bob gets 400,000 tokens (`min((1,000,000 - 600,000), 500,000) = 400,000`), David gets 0 token. The settlement price is **0.000002** ether. Alice collects 0.000002 * 1,000,000 = 2 ethers.

## Judging criteria

- **Smart Contract:** The smart contracts need to be written in Solidity running on the **fhEVM with the Sepolia coprocessor**. The code should be cleanly commented, with adherence (as close as possible) to the standard is important.
- **Testing:** Testing smart contracts using hardhat is also highly recommended.
- **Deployment:** The smart contract should be compatible with Sepolia (using the coprocessor). 
- **Use of fhEVM:** The contract should employ hidden information, showcasing the capabilities of the fhEVM. Care should be taken by the bounty hunter to decide which state variables to encrypt and which should stay unencrypted. Suggestions (non-exhaustive) of state variables to encrypt:  transfer amounts, allowances, number of frozen tokens, identities such as country codes, etc.
- **Bonus points:** Customization of the assets to be sold, a factory-model to create auctions, a small front-end using React (or another framework of your choice!).

### 🥇Best submission: up to $5,000
To be considered best submission, a solution must be efficient, effective and demonstrate a deep understanding of the core problem. Alongside the technical correctness, it should also be submitted with a clean code, clear explanations and a complete documentation.

### 🥈Second-best submission: up to $3,000
For a solution to be considered the second best submission, it should be both efficient and effective. The code should be neat and readable, while its documentation might not be as exhaustive as the best submission, it should cover the key aspects of the solution.

### 🥉Third-best submission: up to $2,000
The third best submission is one that presents a solution that effectively tackles the challenge at hand, even if it may have certain areas of improvement in terms of efficiency or depth of understanding. Documentation should be present, covering the essential components of the solution.

### Step 1: Registration
Click [here](https://www.zama.ai/bounty-program/register) to register for the fhEVM Bounty. Fill out the registration form with your information. Once you fill out the form, you will receive a confirmation email with a link to the submission portal for when you are ready to submit your code.
>[!Note]
>Check your spam folder in case you don't receive the confirmation email. If you haven't received it within 24 hour, please contact us by email at bounty@zama.ai.

### Step 2: Work on the Challenge
Read through the Bounty details and requirements carefully. Use the provided resources and create your own GitHub repository to store your code. 
If you have any questions during your work, feel free to comment directly in the Bounty issue and our team will be happy to assist you.

### Step 3: Submission
Once you have completed your work, upload your completed work to the submission portal using the link provided in the confirmation email.

>[!Note]
>The deadline for submission is February, 9th 2025 (23:59, Anywhere On Earth). Late submissions will not be considered.

We wish you the best of luck with the challenge!

## ✅ Support
- Comment on this issue with any questions regarding this bounty.
- Email for private questions: bounty@zama.ai
- Join the Zama community channels [here](https://zama.ai/community).

**Comment by El-hacen21** [comment]
Hello,

I am considering selling the good (e.g., ERC20) using encrypted ERC20 (ConfidentialERC20).

My motivation is that ConfidentialERC20 would allow to hide the amounts of locked deposits from bidders, as well as the final payment amounts at the end of the auction.

Would this approach be acceptable/preferable? Or is it mandatory to sell "using either Ether or an ERC20 token"? I understand the practical aspects of the latter option, but I wanted to confirm.

Thank you.

**Comment by gabrielstoica** [comment]
Hi @zaccherinij,

There's a minor mistake in the example scenario. The settlement price will be **0.000002** ether, as Bob's bid price is the one at which the last token is bought. Therefore, Alice collects 0.000002 * 1,000,000 = 2 ethers which is now correct.

**Comment by immortal-tofu** [comment]
> is it correct that the original offer and the outcome of the auction can be public?

What is the "original offer"?

The ideal solution would reveal only the settlement price after the auction ends, while keeping the user bids (quantity and price) encrypted. If it is not feasible, the developer should carefully decrypt only what is necessary, avoiding excessive information leakage and defeating the purpose.

**Comment by matthiasgeihs** [comment]
> > is it correct that the original offer and the outcome of the auction can be public?
> 
> What is the "original offer"?

I was referring to the sell offer. (i.e., in the example that would be Alice wants to sell 1,000,000 tokens for ether.)
I'm assuming that all of this would be public information, correct?

> The ideal solution would reveal only the settlement price after the auction ends, while keeping the user bids (quantity and price) encrypted. If it is not feasible, the developer should carefully decrypt only what is necessary, avoiding excessive information leakage and defeating the purpose.

I think some trade offs might be possible if the bids only need to be kept private while the auction is running. E.g., we could just store the encrypted amounts until the auction is over and then decide the outcome, using the bids more as a commitment, than really computing on them. This way we could even make the auction work with standard (non-confidential) ERC20 tokens, because we could do the settlement in the clear.

However, as you also suggested, in the context of this bounty it is probably more interesting if the bidding token is a confidential ERC20 and the bids remain hidden.

@immortal-tofu

**Comment by immortal-tofu** [comment]
Of course the sell offer is public: we know the total amount of token to sell (and the ConfidentialERC20 to use to bid).

Otherwise, I think you get the idea 👍

**Comment by QiteBlock** [comment]
Hello, 

I'd like to confirm something. The settlement price in the example is the minimum bid price in all the winning bids ? 
It means that if Bob only bid 300 000 tokens. And David get the last 100 000 tokens. Then the settlement price is : 0.00000000001 ? 

Thanks

**Comment by immortal-tofu** [comment]
> Hello,
> 
> I'd like to confirm something. The settlement price in the example is the minimum bid price in all the winning bids ? It means that if Bob only bid 300 000 tokens. David get the last 100 000 tokens. Then the settlement price is : 0.00000000001 right ?
> 
> Thanks

Exactly. Winning bids are the highest bids within the limit of available tokens, and the settlement price is the lowest price among these winning bids.
